### ISSUE TITLE
Poweroff script addition

### ISSUE THREAD

**Issue by embryo10** [koreader/koreader#3033]
* KOReader version: 2015.11-1054
* Device: Kobo Glo

Last year I made an [addition to the KSM7 poweroff.sh](https://www.mobileread.com/forums/showthread.php?t=274069) to log the uptime of the reader. 
Since Koreader doesn't use this powerof.sh, is there a way to add that modification to its shutdown script?
The extra commands are:
```bash
uptimelog="/adds/kbmenu/log/uptime.txt"
last=$(awk 'END {print $4}' $uptimelog)
bat="$(cat /sys/devices/platform/pmic_battery.1/power_supply/mc13892_bat/capacity)"
echo "$(date +%Y:%m:%d_%H:%M), $(awk -v last="$last" -v bat="$bat" '{printf("%d:%02d:%02d:%02d total: %d bat: %d",($1/60/60/24),($1/60/60%24),($1/60%60),($1%60),($1+last),(bat))}' /proc/uptime)" >> $uptimelog
```
An alternative would be to somehow use the KSM script from inside Koreader for shutdown.
Is that possible?

**Comment by Frenzie** [comment]
Sure, send a PR? :-)

On Jul 27, 2017 21:08, "embryo10" <notifications@github.com> wrote:


   - KOReader version: 2015.11-1054
   - Device: Kobo Glo

Last year I made an addition to the KSM7 poweroff.sh
<https://www.mobileread.com/forums/showthread.php?t=274069> to log the
uptime of the reader.
Since Koreader doesn't use this powerof.sh, is there a way to add that
modification to its shutdown script?
The extra commands are:

uptimelog="/adds/kbmenu/log/uptime.txt"
last=$(awk 'END {print $4}' $uptimelog)
bat="$(cat /sys/devices/platform/pmic_battery.1/power_supply/mc13892_bat/capacity)"echo
"$(date +%Y:%m:%d_%H:%M), $(awk -v last="$last" -v bat="$bat"
'{printf("%d:%02d:%02d:%02d total: %d bat:
%d",($1/60/60/24),($1/60/60%24),($1/60%60),($1%60),($1+last),(bat))}'
/proc/uptime)" >> $uptimelog

An alternative would be to somehow use the KSM script from inside Koreader
for shutdown.
Is that possible?

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/koreader/koreader/issues/3033>, or mute the thread

**Comment by embryo10** [comment]
Yeah..
That means that after the bash that I had to learn (in the very basic form) bash, I have to start learning lua too, just for that measurement. :p
I will too, when I get some time, but I was trying to weasel  my way out of it for now...

**Comment by Frenzie** [comment]
Oh, the power off is in Lua? I'll take a look next week if I remember.

**Comment by embryo10** [comment]
Actually, I don't know.
Just an assumption.
If it's in bash, thinks could be easier (if I get myself to remember the basics I was telling you about before :-/ )

**Comment by Frenzie** [comment]
@embryo10 Something like this:

```diff
diff --git a/frontend/device/kobo/device.lua b/frontend/device/kobo/device.lua
index 35edf5d3..c22800e5 100644
--- a/frontend/device/kobo/device.lua
+++ b/frontend/device/kobo/device.lua
@@ -726,6 +726,26 @@ function Kobo:saveSettings()
 end
 
 function Kobo:powerOff()
+    local f
+    local battery = ""
+    local uptime = ""
+    local uptime_log
+    f = io.open("/sys/devices/platform/pmic_battery.1/power_supply/mc13892_bat/capacity", "r")
+    if f then
+        battery = f:read("*all")
+        f:close()
+    end
+    f = io.open("/proc/uptime", "r")
+    if f then
+        uptime = f:read("*all")
+        f:close()
+    end
+    uptime_log = io.open("./uptime.txt", "a")
+    uptime_log:write(os.date("%Y:%m:%d_%H:%M"))
+    uptime_log:write(" Uptime: "..uptime.." Battery: "..battery)
+    uptime_log:write("\n")
+    uptime_log:flush()
+    uptime_log:close()
     os.execute("poweroff -f")
 end
 
```

I don't quite understand the /proc/uptime format from a quick glance at your script though. There are two values in it?

**Comment by KenMaltby** [comment]
If we could have an .html as the Poweroff "image".  I am still using KSM7 also and when I use the Power off from KOReader's menu the image from my KSM html is written to the screen before KOReader exits and the KSM menu displays.

**Comment by noembryo** [comment]
Well, just got back.. :o(
Oh, well..

What I finally did, and used it for a while to measure the battery (but not anymore) is this:
1.  I created a script named `poweroff.sh` in the KoReader's dir (started from the KSM `poweroff.sh`)
```bash
#!/bin/sh

ksmroot=${ksmroot:-"/adds/kbmenu"}
if [ "$ksmroot" == "/mnt/onboard/.adds/kbmenu" ]; then
  ksmuser="/mnt/onboard/.adds/kbmenu"
else
  ksmuser=${ksmuser:-"/mnt/onboard/.adds/kbmenu_user"}
fi
uptimelog=$ksmroot/log/uptime.txt

last=$(awk 'END {print $4}' $uptimelog)
bat="$(cat /sys/devices/platform/pmic_battery.1/power_supply/mc13892_bat/capacity)"
echo "$(date +%Y:%m:%d_%H:%M), $(awk -v last="$last" -v bat="$bat" '{printf("%d:%02d:%02d:%02d total: %d bat: %d",($1/60/60/24),($1/60/60%24),($1/60%60),($1%60),($1+last),(bat))}' /proc/uptime)" >> $uptimelog

/bin/busybox poweroff
```
2. I changed a line in the `koreader\frontend\device\kobo\device.lua`
```lua
function Kobo:powerOff()
--    os.execute("poweroff -f")
    os.execute("./poweroff.sh")
```
With that, I did run the script every time I powered off and also had the cover of the book as screensaver :o)

> I don't quite understand the /proc/uptime format from a quick glance at your script though. There are two values in it?

I think that creates a string with these 6 values in it `($1/60/60/24),($1/60/60%24),($1/60%60),($1%60),($1+last),(bat))` (the `$1` must be the `$(date +%Y:%m:%d_%H:%M)`)..
.. or I misunderstood the question?

On a side note, @Frenzie since this is solved, should I close it?

**Comment by Frenzie** [comment]
@noembryo Well, we prefer to keep it all in Lua. We specifically migrated away from helper shell scripts (except the one that helps to set up the environment).

If you don't wish to mainline something like this then I suppose so.

**Comment by noembryo** [comment]
Its much better in lua, I agree.
I just wonder if its really useful to anybody else.

From what I can tell from the code (still not started lua, but I will ;o), what you do is write the duration of use and the battery level when powered off.

What I do, is different.
I forgot to say about a `show_uptime.sh` script that I use to read and display the data from the `uptime.txt`
```bash
#!/bin/sh

uptimelog=${uptimelog:-"/mnt/onboard/.adds/kbmenu/log/uptime.txt"}
xup=$(awk '{printf("%d:%02d:%02d:%02d",($1/60/60/24),($1/60/60%24),($1/60%60),($1%60))}' /proc/uptime)
#xidle=$(awk '{printf("%d:%02d:%02d:%02d",($2/60/60/24),($2/60/60%24),($2/60%60),($2%60))}' /proc/uptime)
up_time=$(awk '{print $1}' /proc/uptime)
last=$(awk 'END {print $4}' $uptimelog)
total=$(awk "BEGIN {print $up_time+$last; exit}")
bat="$(cat /sys/devices/platform/pmic_battery.1/power_supply/mc13892_bat/capacity)"
if [ -s $uptimelog ]; then
    bat_start=$(awk '{print $6; exit}' $uptimelog)
else
    bat_start=$bat
fi
total_bat=$(awk "BEGIN {print $bat_start-$bat; exit}")
xtotal=$(awk -v total="$total" 'BEGIN {printf("%d:%02d:%02d:%02d",(total/60/60/24),(total/60/60%24),(total/60%60),(total%60)); exit}')

echo "Uptime: $xup<br>Total time: $xtotal<br>Battery charge: $bat<br>Battery spend: $total_bat"
```
It will show me the "Uptime", "Total time", "Battery charge" and "Battery spend" from the last reset that could be at any level (not just 100%)

**Comment by Frenzie** [comment]
> I just wonder if its really useful to anybody else.

I don't know about keeping a separate log, but I wouldn't mind writing that to the regular KOReader log on shutdown.

**Comment by noembryo** [comment]
"Uptime" is only when the device is not sleeping, is it not?

And afterwards we have to find all the power off entries in the logs and do the calculations.
Also we have to know if the device is recharged (or USB connected) at any time between the power offs.

Its doable but a little PITA :o)
Better than nothing though..

**Comment by poire-z** [comment]
> And afterwards we have to find all the power off entries in the logs and do the calculations.

Well, better not to have to read the log, and just react to events as they happen and keep counters :)

(I have my own hacks to UIManager to log that kind of things, ... but... :)
We already have something that is supposed to do all that, our Battery statistics plugins:
https://github.com/koreader/koreader/blob/master/plugins/batterystat.koplugin/main.lua
So, may be this plugin just need some additional work to log some state to crash.log and/or show better or extended statistics.

**Comment by noembryo** [comment]
Yes, I know about the plugin, but I don't remember why it didn't help me..
One think I remember, is that it gives me the duration in hours only (something like 4586.35) which is hard to parse.

I could check it again, but I think that the reading statistics are broken in my device for many versions now (no info, crashes, etc.)..
I'm a couple of versions behind, so I will update and if there are still problems I'll create an issue.
Or maybe the battery statistics are in a different db..

**Comment by poire-z** [comment]
It's supposed to log some stuff in koreader/settings/batterstat.lua, and probably just to show some generic stats in its own display.
Nobody (and not me) must really care about this plugin, as it has not been updated since a long time.
Just saying that if any stuff of the kind you and I hacked, should be made into KOReader, we'd probably work/extend/rewrite a plugin like this one.

**Comment by Frenzie** [comment]
I'm referring mainly to the fact that there currently isn't a "Commencing shutdown" notice in the log (but that should be device-agnostic). But if you add that I figure a little extra device-specific line saying something about uptime wouldn't hurt and might please some people. That wouldn't be for stats though. ;-)

@poire-z 
> Nobody (and not me) must really care about this plugin, as it has not been updated since a long time.

While I don't, there's plenty of stuff I do care about that hasn't been updated in years because it just works. (Or at least works well enough.)

**Comment by noembryo** [comment]
> Just saying that if any stuff of the kind you and I hacked, should be made into KOReader, we'd probably work/extend/rewrite a plugin like this one.


If only I could.. (this lack of direct code execution is crippling for newbies like me :o)
If i can be of any help, I'll gladly do whatever I can though.

**Comment by Frenzie** [comment]
@noembryo Device-specific code is definitely the hardest to test, but Kobo's pretty decent on the whole. How do you mean lack of direct code execution? For something basic like I wrote [above](https://github.com/koreader/koreader/issues/3033#issuecomment-526628995) you can basically just run it in Lua or LuaJIT on your desktop and even on the device it's not so bad *except* for testing sleep/shutdown. 

Also keep in mind that just because you're testing something you'd like to happen on shutdown doesn't mean you need to shutdown to test it (unless it's actually shutdown itself you're testing).

**Comment by noembryo** [comment]
> For something basic like I wrote above you can basically just run it in Lua or LuaJIT on your desktop and even on the device it's not so bad except for testing sleep/shutdown.

You mean, there is a way to run KoReader  on the desktop and test the code changes (adding a plugin or modifying one).
I think I remember discussing this again and coming to the conclusion that I can't.
Perhaps the problem is that I use Windows 7 or something.

**Comment by Frenzie** [comment]
Not precisely. I mean that you're not necessarily dealing with KOReader at all. Two weeks ago I wrote [this](https://github.com/koreader/koreader-base/pull/969). My first experiments were in plain LuaJIT. (It uses FFI functionality not available in regular Lua.) Once those were successful I moved it into a crude in-KOReader proof of concept as described [here](https://github.com/koreader/koreader/pull/5335#discussion_r321964943).

But also, for most on-device testing it suffices to edit the file and restart KOReader. Thanks to the gesture manager you can do that really quickly. It's quite convenient actually, as long as you don't have to deal with sleep or shutdown. `tail -f` the crash.log output and the rest is just a matter of your FTP client.

**Comment by Frenzie** [comment]
I'll close this for now, but it's still here for reference of course.
